Presto trying to access default table location even when external_location is provided

From https://prestosql.slack.com/archives/CGB0QHWSW/p1592389129168400

Reproduced on Presto 334.

1. Create a schema with default location pointing to some non-existent/inaccessible node. (In my case this was due to a pre-existing schema on AWS Glue).

```sql
CREATE SCHEMA awsdatacatalog.express_dwh WITH (
    location = 'hdfs://some-inaccessible-node:8020/user/hive/warehouse/express_dwh.db'
)
```

2. Run a CTAS for an external table

```sql
CREATE TABLE new_table WITH (
    format = 'PARQUET',
    external_location = 's3://some-bucket-that-exists/new_table/',
    partitioned_by=ARRAY['ad']
) AS SELECT * FROM some.table
```

This causes an `io.prestosql.spi.PrestoException: Failed checking path: hdfs://some-inaccessible-node:8020/user/hive/warehouse/express_dwh.db` with stack trace:

```log
io.prestosql.spi.PrestoException: Failed checking path: hdfs://ip-some-ip-not-in-cluster:8020/user/hive/warehouse/express_dwh.db
	at io.prestosql.plugin.hive.util.HiveWriteUtils.pathExists(HiveWriteUtils.java:486)
	at io.prestosql.plugin.hive.util.HiveWriteUtils.getTableDefaultLocation(HiveWriteUtils.java:469)
	at io.prestosql.plugin.hive.util.HiveWriteUtils.getTableDefaultLocation(HiveWriteUtils.java:457)
	at io.prestosql.plugin.hive.HiveLocationService.forNewTable(HiveLocationService.java:58)
	at io.prestosql.plugin.hive.HiveMetadata.beginCreateTable(HiveMetadata.java:1271)
	at io.prestosql.plugin.hive.HiveMetadata.beginCreateTable(HiveMetadata.java:255)
	at io.prestosql.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.beginCreateTable(ClassLoaderSafeConnectorMetadata.java:366)
	at io.prestosql.metadata.MetadataManager.beginCreateTable(MetadataManager.java:777)
	at io.prestosql.sql.planner.optimizations.BeginTableWrite$Rewriter.createWriterTarget(BeginTableWrite.java:179)
	at io.prestosql.sql.planner.optimizations.BeginTableWrite$Rewriter.visitTableFinish(BeginTableWrite.java:142)
	at io.prestosql.sql.planner.optimizations.BeginTableWrite$Rewriter.visitTableFinish(BeginTableWrite.java:76)
	at io.prestosql.sql.planner.plan.TableFinishNode.accept(TableFinishNode.java:106)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:84)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:73)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Collections$2.tryAdvance(Collections.java:4747)
	at java.base/java.util.Collections$2.forEachRemaining(Collections.java:4755)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:74)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:38)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:22)
	at io.prestosql.sql.planner.plan.PlanVisitor.visitOutput(PlanVisitor.java:49)
	at io.prestosql.sql.planner.plan.OutputNode.accept(OutputNode.java:83)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter.rewriteWith(SimplePlanRewriter.java:32)
	at io.prestosql.sql.planner.optimizations.BeginTableWrite.optimize(BeginTableWrite.java:73)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:198)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:187)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:182)
	at io.prestosql.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:393)
	at io.prestosql.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:381)
	at io.prestosql.execution.SqlQueryExecution.start(SqlQueryExecution.java:336)
	at io.prestosql.execution.SqlQueryManager.createQuery(SqlQueryManager.java:237)
	at io.prestosql.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:132)
	at io.prestosql.$gen.Presto_334____20200617_013735_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.apache.hadoop.net.ConnectTimeoutException: Call From ip-master/ip-master to ip-some-ip-not-in-cluster:8020 failed on socket timeout exception: org.apache.hadoop.net.ConnectTimeoutException: 500 millis timeout while waiting for channel to be ready for connect. ch : java.nio.channels.SocketChannel[connection-pending remote=ip-some-ip-not-in-cluster:8020]; For more details see:  http://wiki.apache.org/hadoop/SocketTimeout
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at org.apache.hadoop.net.NetUtils.wrapWithMessage(NetUtils.java:831)
	at org.apache.hadoop.net.NetUtils.wrapException(NetUtils.java:775)
	at org.apache.hadoop.ipc.Client.getRpcResponse(Client.java:1515)
	at org.apache.hadoop.ipc.Client.call(Client.java:1457)
	at org.apache.hadoop.ipc.Client.call(Client.java:1367)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:228)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:116)
	at com.sun.proxy.$Proxy294.getFileInfo(Unknown Source)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.getFileInfo(ClientNamenodeProtocolTranslatorPB.java:903)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:422)
	at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeMethod(RetryInvocationHandler.java:165)
	at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invoke(RetryInvocationHandler.java:157)
	at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeOnce(RetryInvocationHandler.java:95)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:359)
	at com.sun.proxy.$Proxy295.getFileInfo(Unknown Source)
	at org.apache.hadoop.hdfs.DFSClient.getFileInfo(DFSClient.java:1665)
	at org.apache.hadoop.hdfs.DistributedFileSystem$29.doCall(DistributedFileSystem.java:1582)
	at org.apache.hadoop.hdfs.DistributedFileSystem$29.doCall(DistributedFileSystem.java:1579)
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
	at org.apache.hadoop.hdfs.DistributedFileSystem.getFileStatus(DistributedFileSystem.java:1594)
	at org.apache.hadoop.fs.FilterFileSystem.getFileStatus(FilterFileSystem.java:454)
	at org.apache.hadoop.fs.FileSystem.exists(FileSystem.java:1683)
	at io.prestosql.plugin.hive.util.HiveWriteUtils.pathExists(HiveWriteUtils.java:483)
	... 40 more
Caused by: org.apache.hadoop.net.ConnectTimeoutException: 500 millis timeout while waiting for channel to be ready for connect. ch : java.nio.channels.SocketChannel[connection-pending remote=ip-172-20-28-16.ec2.internal/172.20.28.16:8020]
	at org.apache.hadoop.net.NetUtils.connect(NetUtils.java:534)
	at org.apache.hadoop.ipc.Client$Connection.setupConnection(Client.java:690)
	at org.apache.hadoop.ipc.Client$Connection.setupIOstreams(Client.java:794)
	at org.apache.hadoop.ipc.Client$Connection.access$3700(Client.java:411)
	at org.apache.hadoop.ipc.Client.getConnection(Client.java:1572)
	at org.apache.hadoop.ipc.Client.call(Client.java:1403)
	... 63 more
```

**NOTE: CTAS works properly in a schema with the proper `location` set**

--- a/presto-hive/src/main/java/io/prestosql/plugin/hive/HiveLocationService.java
+++ b/presto-hive/src/main/java/io/prestosql/plugin/hive/HiveLocationService.java
@@ -55,7 +55,7 @@ public class HiveLocationService
     public LocationHandle forNewTable(SemiTransactionalHiveMetastore metastore, ConnectorSession session, String schemaName, String tableName, Optional<Path> externalLocation)
     {
         HdfsContext context = new HdfsContext(session, schemaName, tableName);
-        Path targetPath = externalLocation.orElse(getTableDefaultLocation(context, metastore, hdfsEnvironment, schemaName, tableName));
+        Path targetPath = externalLocation.orElseGet(() -> getTableDefaultLocation(context, metastore, hdfsEnvironment, schemaName, tableName));
 
         // verify the target directory for the table
         if (pathExists(context, hdfsEnvironment, targetPath)) {

--- a/presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/common/Hadoop.java
+++ b/presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/common/Hadoop.java
@@ -33,6 +33,7 @@ public final class Hadoop
         implements EnvironmentExtender
 {
     public static final String CONTAINER_PRESTO_HIVE_PROPERTIES = CONTAINER_PRESTO_ETC + "/catalog/hive.properties";
+    public static final String CONTAINER_PRESTO_HIVE_WITH_EXTERNAL_WRITES_PROPERTIES = CONTAINER_PRESTO_ETC + "/catalog/hive_with_external_writes.properties";
     public static final String CONTAINER_PRESTO_ICEBERG_PROPERTIES = CONTAINER_PRESTO_ETC + "/catalog/iceberg.properties";
 
     private final DockerFiles dockerFiles;
@@ -62,6 +63,7 @@ public final class Hadoop
         builder.configureContainer("presto-master", container -> {
             container
                     .withCopyFileToContainer(forHostPath(dockerFiles.getDockerFilesHostPath("common/hadoop/hive.properties")), CONTAINER_PRESTO_HIVE_PROPERTIES)
+                    .withCopyFileToContainer(forHostPath(dockerFiles.getDockerFilesHostPath("common/hadoop/hive_with_external_writes.properties")), CONTAINER_PRESTO_HIVE_WITH_EXTERNAL_WRITES_PROPERTIES)
                     .withCopyFileToContainer(forHostPath(dockerFiles.getDockerFilesHostPath("common/hadoop/iceberg.properties")), CONTAINER_PRESTO_ICEBERG_PROPERTIES);
 
             if (System.getenv("HADOOP_PRESTO_INIT_SCRIPT") != null) {

--- a/presto-product-tests-launcher/src/main/resources/docker/presto-product-tests/common/hadoop/hive_with_external_writes.properties
+++ b/presto-product-tests-launcher/src/main/resources/docker/presto-product-tests/common/hadoop/hive_with_external_writes.properties
@@ -1,0 +1,16 @@
+connector.name=hive-hadoop2
+hive.metastore.uri=thrift://hadoop-master:9083
+hive.config.resources=/docker/presto-product-tests/conf/presto/etc/hive-default-fs-site.xml
+hive.allow-add-column=true
+hive.allow-drop-column=true
+hive.allow-rename-column=true
+hive.allow-comment-table=true
+hive.allow-comment-column=true
+hive.allow-drop-table=true
+hive.allow-rename-table=true
+hive.allow-register-partition-procedure=true
+hive.metastore-cache-ttl=0s
+hive.fs.cache.max-size=10
+hive.max-partitions-per-scan=100
+hive.translate-hive-views=true
+hive.non-managed-table-writes-enabled=true

--- a/presto-product-tests/bin/product-tests-suite-2.sh
+++ b/presto-product-tests/bin/product-tests-suite-2.sh
@@ -27,5 +27,10 @@ presto-product-tests-launcher/bin/run-launcher test run \
     -- -g storage_formats,cli,hdfs_impersonation,authorization,hive_file_header -x iceberg,"${DISTRO_SKIP_GROUP}" -e "${DISTRO_SKIP_TEST}" \
     || suite_exit_code=1
 
+presto-product-tests-launcher/bin/run-launcher test run \
+    --environment singlenode \
+    -- -g hive_with_external_writes -x "${DISTRO_SKIP_GROUP}" -e "${DISTRO_SKIP_TEST}" \
+    || suite_exit_code=1
+
 echo "$0: exiting with ${suite_exit_code}"
 exit "${suite_exit_code}"

--- a/presto-product-tests/src/main/java/io/prestosql/tests/TestGroups.java
+++ b/presto-product-tests/src/main/java/io/prestosql/tests/TestGroups.java
@@ -42,6 +42,7 @@ public final class TestGroups
     public static final String HIVE_TRANSACTIONAL = "hive_transactional";
     public static final String HIVE_VIEWS = "hive_views";
     public static final String HIVE_CACHING = "hive_caching";
+    public static final String HIVE_WITH_EXTERNAL_WRITES = "hive_with_external_writes";
     public static final String AUTHORIZATION = "authorization";
     public static final String HIVE_COERCION = "hive_coercion";
     public static final String CASSANDRA = "cassandra";

--- a/presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCreateExternalTable.java
+++ b/presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCreateExternalTable.java
@@ -1,0 +1,51 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.tests.hive;
+
+import com.google.inject.Inject;
+import io.prestosql.tempto.ProductTest;
+import io.prestosql.tempto.hadoop.hdfs.HdfsClient;
+import org.testng.annotations.Test;
+
+import static io.prestosql.tempto.query.QueryExecutor.query;
+import static io.prestosql.tests.TestGroups.HIVE_WITH_EXTERNAL_WRITES;
+import static io.prestosql.tests.TestGroups.PROFILE_SPECIFIC_TESTS;
+import static java.lang.String.format;
+
+public class TestHiveCreateExternalTable
+        extends ProductTest
+{
+    private static final String HIVE_CATALOG_NAME = "hive_with_external_writes";
+
+    @Inject
+    private HdfsClient hdfsClient;
+
+    @Test(groups = {HIVE_WITH_EXTERNAL_WRITES, PROFILE_SPECIFIC_TESTS})
+    public void testCreateExternalTableWithInaccessibleSchemaLocation()
+    {
+        String schema = "schema_without_location";
+        String schemaLocation = "/tmp/" + schema;
+        hdfsClient.createDirectory(schemaLocation);
+        query(format("CREATE SCHEMA %s.%s WITH (location='%s')",
+                HIVE_CATALOG_NAME, schema, schemaLocation));
+
+        hdfsClient.delete(schemaLocation);
+
+        String table = "test_create_external";
+        String tableLocation = "/tmp/" + table;
+        query(format("CREATE TABLE %s.%s.%s WITH (external_location = '%s') AS " +
+                        "SELECT * FROM tpch.tiny.nation",
+                HIVE_CATALOG_NAME, schema, table, tableLocation));
+    }
+}

